tools/automation/cli_linter/linter.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import os
+import json
+import inspect
+import argparse
+from importlib import import_module
+from pkgutil import iter_modules
+
+
+class Linter():
+    def __init__(self, command_table=None, help_file_entries=None, loaded_help=None):
+        self._command_table = command_table
+        self._all_yaml_help = help_file_entries
+        self._loaded_help = loaded_help
+        self._commands = set(command_table.keys())
+        self._command_groups = set()
+        self._parameters = {}
+        self._help_file_entries = set(help_file_entries.keys())
+        self._exit_code = 0
+
+        # get all unsupressed parameters
+        for command_name, command in command_table.items():
+            self._parameters[command_name] = set()
+            for name, param in command.arguments.items():
+                if param.type.settings.get('help') != argparse.SUPPRESS:
+                    self._parameters[command_name].add(name)
+
+        # populate command groups
+        for command_name in self._commands:
+            command_args = []
+            for arg in command_name.split()[:-1]:
+                command_args.append(arg)
+                if command_args:
+                    self._command_groups.add(' '.join(command_args))
+
+    def get_commands(self):
+        return self._commands
+
+    def get_command_groups(self):
+        return self._command_groups
+
+    def get_help_file_entries(self):
+        return self._help_file_entries
+
+    def get_command_parameters(self, command_name):
+        return self._parameters.get(command_name)
+
+    def get_help_entry_type(self, entry_name):
+        return self._all_yaml_help.get(entry_name).get('type')
+
+    def get_command_help(self, command_name):
+        return self._get_loaded_help_description(command_name)
+
+    def get_command_group_help(self, command_group_name):
+        return self._get_loaded_help_description(command_group_name)
+
+    def get_parameter_help(self, command_name, parameter_name):
+        options = self._command_table.get(command_name).arguments.get(parameter_name).type.settings.get('options_list')
+        parameter_helps = self._loaded_help.get(command_name).parameters
+        param_help = next((param for param in parameter_helps if _share_element(options, param.name.split())), None)
+        # workaround for --ids which is not does not generate doc help (BUG)
+        if not param_help:
+            return self._command_table.get(command_name).arguments.get(parameter_name).type.settings.get('help')
+        return param_help.short_summary or param_help.long_summary
+
+    def mark_rule_failure(self):
+        self._exit_code = 1
+
+    def get_exit_code(self):
+        return self._exit_code
```

Comment:
In Python, you don't need all these `get_X(self)` methods.
You can do this:
```
@property
def exit_code(self):
    return self._exit_code
```

At least this is fine for the ones that only have `self` as an argument.